### PR TITLE
use subtest for table units (pkg/scheduler/core)

### LINKED ISSUE #63267: scheduler: Update tests to use subtest
Go 1.7 added the [subtest feature](https://golang.org/doc/go1.7#testing) which can make table-driven tests much easier to run and debug. Many table-driven tests in `pkg/scheduler` and `test/integration/scheduler` are not using this feature.

As an example, https://github.com/kubernetes/kubernetes/pull/56940 converts some tests to use a sub-benchmark.

Further reading: [Using Subtests and Sub-benchmarks](https://blog.golang.org/subtests)

/kind cleanup
/sig scheduling

### PR DESCRIPTION
**Pull request by xchapter7x**
**What this PR does / why we need it**: Update scheduler's unit table tests to use subtest

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:
breaks up PR: https://github.com/kubernetes/kubernetes/pull/63281
/ref #63267

**Release note**:

```release-note
This PR will leverage subtests on the existing table tests for the scheduler units.
Some refactoring of error/status messages and functions to align with new approach.

```

### CODE REVIEW

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 1]
```diff
@@ -692,13 +746,15 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
...
-			if _, exist := algorithmCache.predicatesCache.Get(testPredicate); exist {
-				t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
-					testPredicate, test.nodeName)
-				break
+		t.Run(test.nodeName, func(t *testing.T) {
+			if algorithmCache, exist := ecache.algorithmCache[test.nodeName]; exist {
+				if _, exist := algorithmCache.predicatesCache.Get(testPredicate); exist {
+					t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
+						testPredicate, test.nodeName)
+					return
```
Comment (misterikkit) [inline pkg/scheduler/core/equivalence_cache_test.go]: This won't exit the loop like the break did previously. I think it should be fine to run to the end of the loop though.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 2]
```diff
@@ -692,13 +746,15 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
 
 	// there should be no cached predicate any more
 	for _, test := range tests {
```
Comment (misterikkit) [inline pkg/scheduler/core/equivalence_cache_test.go]: Taking a closer look at this test, I don't like that the various test cases seem to leak state into each other (i.e. loop over the tests twice with the same ecache.)

It's definitely beyond the scope of this PR, but I just wanted to point it out. I should file a separate issue.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 3]
```diff
@@ -156,31 +158,33 @@ func TestSelectHost(t *testing.T) {
...
-					t.Error("Unexpected non-error")
-				}
-			} else {
-				if err != nil {
-					t.Errorf("Unexpected error: %v", err)
-				}
-				if !test.possibleHosts.Has(got) {
-					t.Errorf("got %s is not in the possible map %v", got, test.possibleHosts)
+		t.Run(test.name, func(t *testing.T) {
+			// increase the randomness
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler_test.go]: More commentary on the pre-existing tests: This isn't super random...

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 4]
```diff
@@ -396,39 +400,41 @@ func TestGenericScheduler(t *testing.T) {
...
+				algorithm.EmptyPriorityMetadataProducer,
+				[]algorithm.SchedulerExtender{},
+				nil,
+				pvcLister,
+				test.alwaysCheckAllPredicates,
+				false)
+			machine, err := scheduler.Schedule(test.pod, schedulertesting.FakeNodeLister(makeNodeList(test.nodes)))
+
+			if !reflect.DeepEqual(err, test.wErr) {
+				t.Errorf("Failed : Unexpected error: %v, expected: %v", err, test.wErr)
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler_test.go]: Can also remove the "Failed :" prefix from these messages.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 5]
```diff
@@ -629,55 +639,61 @@ func TestZeroRequest(t *testing.T) {
...
-		mataData := mataDataProducer(test.pod, nodeNameToInfo)
-
-		list, err := PrioritizeNodes(
-			test.pod, nodeNameToInfo, mataData, priorityConfigs,
-			schedulertesting.FakeNodeLister(test.nodes), []algorithm.SchedulerExtender{})
-		if err != nil {
-			t.Errorf("unexpected error: %v", err)
-		}
-		for _, hp := range list {
-			if test.test == "test priority of larger pod with machine with zero-request pod" {
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler_test.go]: I don't know what to say about this...

**Review by misterikkit — COMMENTED** [review]
It looks like you pulled some test bodies into helper functions and left others inline. My guess is that you did this to improve readability, but I feel that keeping them all inline is more readable. Even though it's a lot of code, it only makes sense in the context of the one test where it is used.

This is clearly subjective, so I'm open to be convinced otherwise. What is the threshold for a test size before it should be pulled into a helper?

**Comment by misterikkit** [comment]
> It looks like you pulled some test bodies into helper functions and left others inline.
Since this pattern appears in a few of your PRs, I'd like to resolve it first. How do you feel about keeping tests inline?

**Comment by xchapter7x** [comment]
@misterikkit , i think the discussion to break out the tests or to keep them inline can be deferred.
Thinking about it now, it feels out of scope to this effort.

So, I agree, lets keep them inline for now. 
let me re-work this PR. 
im gonna put this one back to WIP for now. stay tuned.

**Comment by xchapter7x** [comment]
/assign @misterikkit
this PR should be in a good spot if you have a moment to take a look. thanks.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 6]
```diff
@@ -342,40 +342,42 @@ func TestUpdateResult(t *testing.T) {
...
+				test.fit,
+				test.reasons,
+				test.equivalenceHash,
+				test.cache,
+				node,
+			)
+
+			cachedMapItem, ok := ecache.algorithmCache[test.nodeName][test.predicateKey]
+			if !ok {
+				t.Errorf("Failed: %s, can't find expected cache item: %v",
```
Comment (misterikkit) [inline pkg/scheduler/core/equivalence_cache_test.go]: Remove test.name  and "Failed" from log messages. I think you had done this in a previous version, and accidentally reverted it?

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 7]
```diff
@@ -689,31 +697,40 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
...
-		ecache.updateResult(
-			test.podName,
-			testPredicate,
-			test.cachedItem.fit,
-			test.cachedItem.reasons,
-			test.equivalenceHashForUpdatePredicate,
-			test.cache,
-			node,
-		)
+		t.Run(test.name, func(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/core/equivalence_cache_test.go]: Given the nature of this test, I think we should actually run this first loop in the main test, and use t.Run in the second loop like you are already doing.

My reasoning is that this test uses the same ecache object across all test cases. (I think that should be fixed up in a separate PR.) so this loop feels more like setup than the actual test.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 8]
```diff
@@ -708,12 +714,19 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
...
-				t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
-					testPredicate, test.nodeName)
-				break
+		breakOnTestFailure := false
+		t.Run(test.name, func(t *testing.T) {
+			if algorithmCache, exist := ecache.algorithmCache[test.nodeName]; exist {
+				if _, exist := algorithmCache[testPredicate]; exist {
+					t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
+						testPredicate, test.nodeName)
+					breakOnTestFailure = true
```
Comment (bsalamat) [inline pkg/scheduler/core/equivalence_cache_test.go]: I think we can remove this and continue testing other cases when one fails.

File: pkg/scheduler/core/equivalence_cache_test.go (hunk 7 above)
Comment (bsalamat) [reply pkg/scheduler/core/equivalence_cache_test.go]: I agree with @misterikkit. You shouldn't `t.Run` in this loop.

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 5 above)
Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler_test.go]: This is really ugly, but it existed. We should consider fixing this in a separate PR.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 9]
```diff
@@ -476,51 +476,53 @@ func TestLookupResult(t *testing.T) {
...
+			// calculate predicate with equivalence cache
+			fit, reasons, invalid := ecache.lookupResult(test.podName,
+				test.nodeName,
+				test.predicateKey,
+				test.equivalenceHashForCalPredicate,
+			)
+			// returned invalid should match expectedInvalidPredicateKey or expectedInvalidEquivalenceHash
+			if test.equivalenceHashForUpdatePredicate != test.equivalenceHashForCalPredicate {
+				if invalid != test.expectedInvalidEquivalenceHash {
+					t.Errorf("Failed: %s, expected invalid: %v, but got: %v",
```
Comment (bsalamat) [inline pkg/scheduler/core/equivalence_cache_test.go]: To be consistent, please remove `test.name` here.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 10]
```diff
@@ -476,51 +476,53 @@ func TestLookupResult(t *testing.T) {
...
+						test.name, test.expectedInvalidEquivalenceHash, invalid)
+				}
+			} else {
+				if invalid != test.expectedInvalidPredicateKey {
+					t.Errorf("Failed: %s, expected invalid: %v, but got: %v",
```
Comment (bsalamat) [inline pkg/scheduler/core/equivalence_cache_test.go]: ditto

**Review by bsalamat — COMMENTED** [review]
Thanks, @xchapter7x!

**Comment by xchapter7x** [comment]
@bsalamat thanks for the feedback. 

changes are in and merge conflicts should also be resolved.
Just waiting for all checks to go green.

- John

**Review by bsalamat — APPROVED** [review]
/lgtm

Thanks, @xchapter7x!
